control.lua

```diff
@@ -156,165 +158,205 @@ end
+	pole.destructible = false
+	pole.disconnect_neighbour()
+	pole.connect_neighbour(poles[#poles])
+	table.insert(poles, pole)
+	return pole
 end
--- For update 11
-function update_all_power_settings()
-	for _, factory in pairs(global.factories) do
-		update_power_settings(factory)
+
+local function connect_power(factory, pole)
+	if not reduce_neighbours(pole) then return end
+	factory.outside_power_pole = pole
+	
+	if factory.inside_surface.name ~= pole.surface.name then
+		available_pole(factory).connect_neighbour(pole)
+		factory.direct_connection = pole
+		return
 	end
+	
+	local n
+	for i, pole in ipairs(global.middleman_power_poles) do
+		if pole == 0 then n = i break end
+	end
+	n = n or #global.middleman_power_poles + 1
+	
+	local surface = power_middleman_surface()
+	local middleman = surface.create_entity{name = "factory-power-connection", position = {2*(n%32), 2*math.floor(n/32)}, force = "neutral"}
+	middleman.destructible = false
+	global.middleman_power_poles[n] = middleman
+	
+	middleman.connect_neighbour(available_pole(factory))
+	middleman.connect_neighbour(pole)
+	
+	factory.middleman_id = n
 end
 
-local function adjust_power_transfer_rate(factory, positive)
-	local transfer_rate = factory.transfer_rate
-	if positive then
-		for i = 1,#VALID_POWER_TRANSFER_RATES do
-			if transfer_rate < VALID_POWER_TRANSFER_RATES[i] then
-				transfer_rate = VALID_POWER_TRANSFER_RATES[i]
-				break
-			end
-		end
-		if transfer_rate > VALID_POWER_TRANSFER_RATES[#VALID_POWER_TRANSFER_RATES] then
-			transfer_rate = VALID_POWER_TRANSFER_RATES[#VALID_POWER_TRANSFER_RATES]
-		end
-	else
-		for i = #VALID_POWER_TRANSFER_RATES,1,-1 do
-			if transfer_rate > VALID_POWER_TRANSFER_RATES[i] then
-				transfer_rate = VALID_POWER_TRANSFER_RATES[i]
-				break
-			end
+function update_power_connection(factory, pole) -- pole parameter is optional
+	local electric_network = factory.outside_energy_receiver.electric_network_id
+	if electric_network == nil then return end
+	local surface = factory.outside_surface
+	local x = factory.outside_x
+	local y = factory.outside_y
+	
+	if not script.active_mods['factorissimo-power-pole-addon'] and global.surface_factory_counters[surface.name] then
+		local surrounding = find_surrounding_factory(surface, {x=x, y=y})
+		if surrounding then
+			connect_power(factory, available_pole(surrounding))
+			return
 		end
-		if transfer_rate < VALID_POWER_TRANSFER_RATES[1] then
-			transfer_rate = VALID_POWER_TRANSFER_RATES[1]
+	end
+	
+	-- find the nearest connected power pole
+	local D = game.max_electric_pole_supply_area_distance + factory.layout.outside_size / 2
+	local canidates = {}
+	for _, entity in ipairs(surface.find_entities_filtered{type="electric-pole", area={{x-D, y-D}, {x+D,y+D}}}) do
+		if entity.electric_network_id == electric_network and entity ~= pole then
+			canidates[#canidates+1] = entity
```

Comment:
You can use the index from the loop (`_` in this case, or rename it to `k`),
That way it won't have to recalculate the size every addition

Also then you must loop with `pairs`, since it will be sparse

But it is recommended to always loop with `pairs` in factorio, as performance is the same and it handles all the cases better